test/runtests.jl

```diff
@@ -258,6 +263,23 @@ const PyInt = pyversion < v"3" ? Int : Clonglong
     end
     @test convert(BigInt, PyObject(1234)) == 1234
 
+    # hasproperty, getproperty, and propertynames
+    py"""
+    class A:
+        class B:
+            C = 1
+    """
+    A = py"A"
+    @test hasproperty(A, "B")
+    @test getproperty(A, "B") == py"A.B"
+    @test :B in propertynames(A)
+    @static if VERSION >= v"0.7-"
+        @test A.B.C == 1
+        @test_throws KeyError A.X
```

Comment:
It's nice that it's tested, but now that we write it as property access, isn't it a bit funny to throw KeyError?  But Julia doesn't have "UndefPropertyError"...  getindex seems to throw ErrorException but that's not super nice.  Still, I guess the best choice here is ErrorException as it would correctly communicate to the users to not rely on the exception type.

Reply:
Maybe `UndefRefError` is OK to use?:

```julia
help?> UndefRefError
search: UndefRefError

  UndefRefError()

  The item or field is not defined for the given object.
```

(Julia has clear distinction between field and property so it may be misleading tough...)

It can be invoked by

```julia
julia> struct UndefFields
           a
           UndefFields() = new()
       end

julia> UndefField()
UndefFields(#undef)

julia> err = try UndefFields().a; catch err; err; end
UndefRefError()
```

What I was referring earlier was this behavior:

```julia
julia> err = try UndefFields().b; catch err; err; end
ErrorException("type UndefFields has no field b")
```